client/src/communication/udp.rs

```diff
@@ -0,0 +1,563 @@
+            request_header_length: (buf[UDPMessageHeader::MSG_HEADER_LEN_MSB] as u16) << 8 |
+                buf[UDPMessageHeader::MSG_HEADER_LEN_LSB] as u16,
+        };
+        Ok(header)
+    }
+
+    /// Check equality with another UDPMessageHeader
+    ///
+    /// # Returns
+    /// True if equal, otherwise false
+    #[cfg(test)]
+    fn equal(&mut self, header: &UDPMessageHeader) -> bool {
+        self.request_header_length == header.request_header_length &&
+            self.body_length == header.body_length &&
+            self.url_length == header.url_length &&
+            self.status_code == header.status_code &&
+            self.message_length == header.message_length
+    }
+}
+
+/// Initialise a UDP socket
+///
+/// # Returns
+/// A UDP socket bound to given local address and port
+fn init_host(local_host: String, local_port: String, timeout: u64) -> net::UdpSocket {
+    let host = format!("{}:{}", local_host, local_port);
+    let socket = net::UdpSocket::bind(host).expect("Failed to init UDP socket");
+    socket.set_read_timeout(Some(std::time::Duration::from_secs(timeout)))
+        .expect("Failed to set timeout");
+    socket
+}
+
+/// Sends a message over a UDP socket
+fn send(socket: &net::UdpSocket, receiver: String, msg: &Vec<u8>) {
+    match socket.send_to(&msg, receiver) {
+        Ok(_) => {},
+        Err(e) => {
+            #[cfg(debug_assertions)]
+            println!("Failed send message over socket: {}", e);
+        }
+    }
+}
+
+/// Listen to a socket to receive a message
+///
+/// # Returns
+/// Vector of bytes received from the socket. If any error occured, an enpty vector will be
+/// returned.
+fn listen(socket: &net::UdpSocket) -> Vec<u8> {
+    let mut buf: [u8; UDPMessage::MAX_LENGTH as usize] = [0; UDPMessage::MAX_LENGTH as usize];
+    let mut result: Vec<u8> = Vec::new();
+    match socket.recv(&mut buf) {
+        Ok(bytes_read) => {
+            #[cfg(debug_assertions)]
+            println!("Message received, {} bytes", bytes_read);
+            result = Vec::from(&buf[0..bytes_read]);
+        },
+        Err(e) => {
+            #[cfg(debug_assertions)]
+            println!("Failed receive message over socket: {}", e);
+        }
+    };
+    result
+}
+
+fn get_authorization_header(token: &String) -> String {
+    return format!("\"Authorization\": {}", token);
+}
+
+/// Send a post request using the UDP protocol
+///
+/// # Returns
+/// If expecting a response, the reponse body will be returned
+///
+/// # Errors
+/// If expecting response, and no response was returned from remote host or if parsing of the
+/// response fails
+pub fn post_request_udp(
+    body: String,
+    url: String,
+    auth_token: &String,
+    remote_host: String,
+    remote_port: String,
+    expecting_response: bool) -> Result<String, anyhow::Error> {
+    let header = get_authorization_header(auth_token);
+    let mut message = UDPMessage::new(url, header, body, 0);
```

Comment:
Just realised that perhaps we need to use the status codes when sending requests to the server as well. Maybe we should specify that the request we are sending is, for example, a GET or a POST using the `status_code` field. It would be a good way of letting the control server know how what HTTP method should be used when relaying the request to the control servers HTTP API.

Reply:
Fixed in [4120ebf](https://github.com/ghmo2789/puppet-master/pull/103/commits/4120ebfc4845cdb9b902fcc94d5b7d3194e3bf5a)!